### DISCUSSION TITLE
Hazelcast Implementation

### DISCUSSION THREAD

**Discussion by testark** [vladimir-bukhtoyarov/bucket4j#6] [Q&A] [answered]
@vladimir-bukhtoyarov is the provided Hazelcast implementation safe across multiple JVMs? I mean, does it ensure that two nodes creating a bucket simultaneously on a given named `IMap<K, V>` will only actually create one single bucket (without resetting a previously created one with the same key)?

**Comment by vladimir-bukhtoyarov** [comment, marked as answer]
Yes, Hazelcast implementation is safe(same for GridGain and Coherence), it never replaces bucket which already exists:

``` java
public class HazelcastProxy implements GridProxy {

    ...

    @Override
    public void setInitialState(GridBucketState initialState) {
        map.putIfAbsent(key, initialState);
    }

    ...

}
```

This behavior is guaranteed by [putIfAbsent](http://docs.hazelcast.org/docs/latest/javadoc/com/hazelcast/core/IMap.html#putIfAbsent%28K,%20V%29)

**Comment by testark** [comment]
@vladimir-bukhtoyarov that's great :-) Thank you so much! Great project, by the way :-)
